zndraw/frame.py

```diff
@@ -0,0 +1,191 @@
+import copy
+import dataclasses
+import typing as t
+
+import ase
+import networkx as nx
+import numpy as np
+from ase.calculators.singlepoint import SinglePointCalculator
+from ase.data.colors import jmol_colors
+from ase.neighborlist import natural_cutoffs
+from pydantic import Field
+
+from zndraw.bonds import ASEComputeBonds
+from zndraw.data import _get_radius, _rgb2hex
+
+
+@dataclasses.dataclass
+class Frame:
+    # Main Attributes
+    positions: t.Union[np.ndarray, list] = None
+    cell: t.Union[np.ndarray, list] = np.array([0.0, 0.0, 0.0])
+    numbers: t.Union[np.ndarray, list, int] = None
+    colors: t.Union[np.ndarray, list] = None
+    radii: t.Union[np.ndarray, list] = None
+    pbc: bool = False
+    connectivity: nx.Graph() = nx.Graph()
+    calc: t.Any = None
+
+    vecField: t.Any = np.array(["TestArray", "0"])
+    # Secondary Attributes
+    # define calculations in this class
+    bonds: bool = True
+    auto_bonds: bool = True
+
+    def __post_init__(self):
+        if isinstance(self.positions, list):
+            self.positions = np.array(self.positions)
+        if isinstance(self.positions, list):
+            self.cell = np.array(self.cell)
+        if isinstance(self.positions, list):
+            self.numbers = np.array(self.numbers)
+        if isinstance(self.positions, list):
+            self.colors = np.array(self.colors)
+        if isinstance(self.positions, list):
+            self.radii = np.array(self.radii)
+        # i know its ugly, but it works
+        # if i convert them without the check, the speed drops dramatically
+
+    @classmethod
+    def from_atoms(cls, atoms: ase.Atoms):
+        frame = cls(**atoms.arrays)
+
+        if hasattr(atoms, "connectivity"):
+            frame.connectivity = atoms.connectivity
+
+        try:
+            calc_data = {}
+            for key in atoms.calc.results:
+                value = atoms.calc.results[key]
+                if isinstance(value, np.ndarray):
+                    value = value.tolist()
+                calc_data[key] = value
+
+            frame.calc = calc_data
+        except (RuntimeError, AttributeError):
+            pass
```

Comment:
I am not entirely sure, when this happens. This was adopted from the previous code. I assume it had something to do with the object having no calc attribute, but this should not happen anymore as the dictionary is now an empty dict even if the atom has no calc attribute. This could probably be removed